ActionView#content_for uses keyword argument to replace options hash

drjayvee opened the pull request:
# Motivation / Background

This PR improves the signature for `content_for` by replacing the `options` hash with a modern keyword argument.

The benefits are:
* Invalid arguments will eventually (see below) `raise ArgumentError` instead of being silently ignored.
  (e.g., `content_for :foo, fluhs: true { "whoops" }`)
* Keyword arguments are documented in the method's signature instead of in its body.
  Better for humans, IDEs and LLMs.
* Remove confusing argument shifting (`options = content if content` 🤨).

## Update
Based on seanpdoyle's review, the scope of this change was reduced to maintain backwards compatibility.

The only change now is that options other than `flush` will trigger a deprecation warning.

# Backwards-compatibility
The current behavior is kept but deprecated. The added warning will not only warn users of a future change, it will also alert them to bugs in their code. (Any invalid option is almost guaranteed to be a typo!)

~~Note that I've added the BC-keeping code in the second commit. If this is deemed overkill ("progress over stability" 😉), I can easily drop it. Otherwise, I'd be more than happy to squash the two commits.~~

# Checklist

* [X] This Pull Request is related to one change. Unrelated changes should be opened in separate PRs.
* [X] Commit message has a detailed description of what changed and why. If this PR fixes a related issue include it in the commit message. Ex: `[Fix #issue-number]`
* [X] Tests are added or updated if you fix a bug or add a feature.
* [X] CHANGELOG files are updated for the changed libraries if there is a behavior change or additional feature. Minor bug fixes and documentation changes should not be included.

seanpdoyle commented on actionview/lib/action_view/helpers/capture_helper.rb:
```diff
@@ -169,14 +169,20 @@ def capture(*, **, &block)
+          ActionView.deprecator.warn(<<-MSG.squish)
+            Passing invalid options to content_for is deprecated.
+            Options #{options.keys.join(', ')} will be ignored.
+          MSG
```
Is this worth mentioning in the Action View Deprecations section of the [8.2 Release Notes](https://github.com/rails/rails/blob/61b5c7e2953fd0ccfab151b233ce69283c567a34/guides/source/8_2_release_notes.md#L57-L64)?

seanpdoyle replied:
If this change is merged, the next release that it will be part of is 8.2.

In its current form, it will start emitting deprecation warnings. The Release Notes serve as an outward-facing list of behaviors that are deprecated, so adding an entry makes the motivation behind this underlying change more discoverable.

> Also, will this be deprecated in 8.2 and removed in 8.3

If the commit is merged ahead of the 8.2 release, it will be part of the 8.2 release. The timing of its removal (whether 8.3 or later) is up to Core.

drjayvee replied:
All right, I'll add it to the `8_2_release_notes`.

drjayvee committed:
Deprecate invalid options for content_for

Invalid options were previously silently ignored. For example,
`content_for :foo, fluhs: true { "bar" }` (note the typo) would not
flush content, with no indication of the mistake.

In the next major release, unknown options will raise ArgumentError.

The options hash argument is intentionally kept (rather than switching
to a keyword argument) to preserve backwards compatibility with callers
that pass an explicit Hash, e.g.:

  opts = { flush: true }
  content_for :foo, opts do "bar" end

Switching to **options would silently change behaviour for these callers
(the hash would land in content instead). That migration can happen in a
future major release once callers have been notified via deprecations.

Eventually, the signature should become

  def content_for(name, content = nil, flush: false, &block)

Not only does that clarify which keyword arguments are supported, the
argument shifting (`options = content if content`) is then no longer
necessary.
